File: Code Generation/responses/python/gemini/medium/closest-equal-element-queries.py
from typing import List
from collections import defaultdict
import bisect

class Solution:
    def solveQueries(self, nums: List[int], queries: List[int]) -> List[int]:
        n = len(nums)
        # Map each value to a sorted list of indices where it appears
        val_to_indices = defaultdict(list)
        for i, val in enumerate(nums):
            val_to_indices[val].append(i)
            
        results = []
        for q_idx in queries:
            val = nums[q_idx]
            indices = val_to_indices[val]
            
            # If the element appears only once, there is no "other" index j
            if len(indices) <= 1:
                results.append(-1)
                continue
            
            # Find the position of q_idx in the sorted list of indices
            pos = bisect.bisect_left(indices, q_idx)
            
            # Potential neighbors in the sorted list (circularly)
            # Predecessor (left neighbor)
            left_neighbor = indices[(pos - 1) % len(indices)]
            # Successor (right neighbor)
            right_neighbor = indices[(pos + 1) % len(indices)]
            
            # Calculate circular distances
            # Distance = min(|i - j|, n - |i - j|)
            def get_circular_dist(i, j, n):
                linear_dist = abs(i - j)
                return min(linear_dist, n - linear_dist)
            
            dist_left = get_circular_dist(q_idx, left_neighbor, n)
            dist_right = get_circular_dist(q_idx, right_neighbor, n)
            
            results.append(min(dist_left, dist_right))
            
        return results

File: Code Generation/responses/cpp/gemini/medium/closest-equal-element-queries.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <cmath>

using namespace std;

class Solution {
public:
    vector<int> solveQueries(vector<int> nums, vector<int> queries) {
        int n = nums.size();
        unordered_map<int, vector<int>> positions;
        
        // Group all indices by their value
        for (int i = 0; i < n; ++i) {
            positions[nums[i]].push_back(i);
        }
        
        vector<int> answer;
        answer.reserve(queries.size());
        
        for (int q_idx : queries) {
            int val = nums[q_idx];
            const vector<int>& indices = positions[val];
            
            // If the element appears only once, there's no "other" index
            if (indices.size() <= 1) {
                answer.push_back(-1);
                continue;
            }
            
            // Find the position of q_idx within the sorted list of indices for this value
            auto it = lower_bound(indices.begin(), indices.end(), q_idx);
            int idx_in_pos = distance(indices.begin(), it);
            
            int min_dist = n; // Initialize with max possible distance
            
            // The nearest identical elements in a circular array are either:
            // 1. The immediate neighbor in the sorted indices list (linear distance)
            // 2. The wrapping neighbor (distance across the array boundary)
            
            // Check neighbor to the left (circularly)
            int left_idx = (idx_in_pos - 1 + indices.size()) % indices.size();
            int dist_left = abs(q_idx - indices[left_idx]);
            dist_left = min(dist_left, n - dist_left);
            min_dist = min(min_dist, dist_left);
            
            // Check neighbor to the right (circularly)
            int right_idx = (idx_in_pos + 1) % indices.size();
            int dist_right = abs(q_idx - indices[right_idx]);
            dist_right = min(dist_right, n - dist_right);
            min_dist = min(min_dist, dist_right);
            
            answer.push_back(min_dist);
        }
        
        return answer;
    }
};